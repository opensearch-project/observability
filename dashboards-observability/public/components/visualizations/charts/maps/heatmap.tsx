/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */
import React, { useMemo } from 'react';
import { colorPalette } from '@elastic/eui';
import { forEach, has, isEmpty, isEqual, uniq } from 'lodash';
import Plotly from 'plotly.js-dist';
import {
  HEATMAP_PALETTE_COLOR,
  HEATMAP_SINGLE_COLOR,
  OPACITY,
  SINGLE_COLOR_PALETTE,
} from '../../../../../common/constants/colors';
import {
  hexToRgb,
  lightenColor,
  getPropName,
} from '../../../../components/event_analytics/utils/utils';
import { IVisualizationContainerProps } from '../../../../../common/types/explorer';
import { VisCanvassPlaceholder } from '../../../event_analytics/explorer/visualizations/shared_components/vis_canvass_placeholder';
import { Plt } from '../../plotly/plot';
import { AGGREGATIONS, GROUPBY } from '../../../../../common/constants/explorer';
import { PLOT_MARGIN } from '../../../../../common/constants/shared';
import { getCompleteTimespanKey } from '../../../visualizations/charts/shared/common';
import { removeBacktick } from '../../../../../common/utils';

export const HeatMap = ({ visualizations, layout, config }: any) => {
  const {
    data: {
      rawVizData: {
        data: queriedVizData,
        metadata: { fields },
      },
      userConfigs: {
        dataConfig: {
          chartStyles = {},
          span = {},
          legend = {},
          tooltipOptions = {},
          panelOptions = {},
          [GROUPBY]: dimensions = [],
          [AGGREGATIONS]: series = [],
        } = {},
        layoutConfig = {},
      } = {},
    } = {},
    vis: { icontype },
  }: IVisualizationContainerProps = visualizations;

  const backtickRemovedVisData = {};
  forEach(queriedVizData, (value, key) => {
    backtickRemovedVisData[removeBacktick(key)] = value;
  });

  const combinedDemensions = [
    ...(!isEmpty(span) ? [getCompleteTimespanKey(span)] : []),
    ...dimensions,
  ];
  if (!isEqual(combinedDemensions.length, 2) || !isEqual(series.length, 1))
    return <VisCanvassPlaceholder message="Invalid data, heatmap can only have exact 2 dimensions and 1 series" icon={icontype} />;
  const xaxisField = { ...combinedDemensions[0] };
  const yaxisField = { ...combinedDemensions[1] };
  const zMetrics = { ...series[0] };
  if (
    isEmpty(xaxisField) ||
    isEmpty(yaxisField) ||
    isEmpty(zMetrics) ||
    isEmpty(backtickRemovedVisData[removeBacktick(xaxisField.label)]) ||
    isEmpty(backtickRemovedVisData[removeBacktick(yaxisField.label)])
  )
    return <VisCanvassPlaceholder message="Invalid data" icon={icontype} />;

  const uniqueYaxis = uniq(backtickRemovedVisData[removeBacktick(yaxisField.label)]);
  const uniqueXaxis = uniq(backtickRemovedVisData[removeBacktick(xaxisField.label)]);
  const uniqueYaxisLength = uniqueYaxis.length;
  const uniqueXaxisLength = uniqueXaxis.length;
  const tooltipMode =
    tooltipOptions.tooltipMode !== undefined ? tooltipOptions.tooltipMode : 'show';
  const tooltipText = tooltipOptions.tooltipText !== undefined ? tooltipOptions.tooltipText : 'all';

  const colorField = chartStyles
    ? chartStyles.colorMode && chartStyles.colorMode[0].name === OPACITY
      ? chartStyles.color ?? HEATMAP_SINGLE_COLOR
      : chartStyles.scheme ?? HEATMAP_PALETTE_COLOR
    : HEATMAP_PALETTE_COLOR;
  const showColorscale = legend.showLegend ?? 'show';

  const traceColor: any = [];
  if (colorField.name === SINGLE_COLOR_PALETTE) {
    const colorsArray = colorPalette([lightenColor(colorField.color, 50), colorField.color], 10);
    colorsArray.map((hexCode, index) => {
      traceColor.push([
        (index !== colorsArray.length - 1 ? index : 10) / 10,
        hexToRgb(hexCode, 1, false),
      ]);
    });
  }

  const calculatedHeapMapZaxis: Plotly.Data[] = useMemo(() => {
    const heapMapZaxis = [];
    const buckets = {};

    // maps bukcets to metrics
    for (let i = 0; i < backtickRemovedVisData[removeBacktick(xaxisField.label)].length; i++) {
      buckets[`${backtickRemovedVisData[removeBacktick(xaxisField.label)][i]},${backtickRemovedVisData[removeBacktick(yaxisField.label)][i]}`] =
      backtickRemovedVisData[getPropName(zMetrics)][i];
    }

    // initialize empty 2 dimensional array, inner loop for each xaxis field, outer loop for yaxis
    for (let i = 0; i < uniqueYaxisLength; i++) {
      const innerBuckets = [];
      for (let j = 0; j < uniqueXaxisLength; j++) {
        innerBuckets.push(null);
      }
      heapMapZaxis.push(innerBuckets);
    }

    // fill in each bucket
    for (let i = 0; i < uniqueYaxisLength; i++) {
      for (let j = 0; j < uniqueXaxisLength; j++) {
        if (has(buckets, `${uniqueXaxis[j]},${uniqueYaxis[i]}`)) {
          heapMapZaxis[i][j] = buckets[`${uniqueXaxis[j]},${uniqueYaxis[i]}`];
        }
      }
    }

    return heapMapZaxis;
  }, [
    backtickRemovedVisData,
    uniqueYaxis,
    uniqueXaxis,
    uniqueYaxisLength,
    uniqueXaxisLength,
    xaxisField,
    yaxisField,
    zMetrics,
  ]);

  const heapMapData = [
    {
      z: calculatedHeapMapZaxis,
      x: uniqueXaxis,
      y: uniqueYaxis,
      hoverinfo: tooltipMode === 'hidden' ? 'none' : tooltipText,
      colorscale: colorField.name === SINGLE_COLOR_PALETTE ? traceColor : colorField.name,
      type: 'heatmap',
      showscale: showColorscale === 'show',
    },
  ];

  layout.yaxis = { autosize: true, automargin: true };
  const mergedLayout = {
    ...layout,
    ...(layoutConfig.layout && layoutConfig.layout),
    title: panelOptions.title || layoutConfig.layout?.title || '',
    margin: PLOT_MARGIN,
  };

  const mergedConfigs = useMemo(
    () => ({
      ...config,
      ...(layoutConfig.config && layoutConfig.config),
    }),
    [config, layoutConfig.config]
  );

  return <Plt data={heapMapData} layout={mergedLayout} config={mergedConfigs} />;
};
