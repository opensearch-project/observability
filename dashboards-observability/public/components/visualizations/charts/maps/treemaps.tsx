/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { isEmpty, isEqual, uniq } from 'lodash';
import React, { useMemo } from 'react';

import {
  DEFAULT_PALETTE,
  MULTI_COLOR_PALETTE,
  SINGLE_COLOR_PALETTE,
} from '../../../../../common/constants/colors';
import { AGGREGATIONS, GROUPBY } from '../../../../../common/constants/explorer';
import { DEFAULT_CHART_STYLES, PLOT_MARGIN } from '../../../../../common/constants/shared';
import { IVisualizationContainerProps } from '../../../../../common/types/explorer';
import { EmptyPlaceholder } from '../../../event_analytics/explorer/visualizations/shared_components/empty_placeholder';
import { Plt } from '../../plotly/plot';

export const TreeMap = ({ visualizations, layout, config }: any) => {
  const { DefaultSortSectors } = DEFAULT_CHART_STYLES;
  const {
    data: {
      defaultAxes,
      indexFields,
      query,
      rawVizData: {
        data: queriedVizData,
        metadata: { fields },
      },
      userConfigs: {
        dataConfig: {
          chartStyles = {},
          legend = {},
          tooltipOptions = {},
          panelOptions = {},
          treemapOptions = {},
          [GROUPBY]: dimensions = [],
          [AGGREGATIONS]: series = [],
        },
        layoutConfig = {},
      },
    },
    vis: { icontype },
  }: IVisualizationContainerProps = visualizations;

  const childField =
    dimensions && dimensions[0]?.childField ? dimensions[0]?.childField : fields[fields.length - 1];
  const parentFields = dimensions && dimensions[0]?.parentFields ? dimensions[0]?.parentFields : [];
  const tooltipMode =
    tooltipOptions.tooltipMode !== undefined ? tooltipOptions.tooltipMode : 'show';
  const tooltipText = tooltipOptions.tooltipText !== undefined ? tooltipOptions.tooltipText : 'all';
  const valueField = series && series[0]?.valueField ? series[0]?.valueField : fields[0];
  const colorField =
    chartStyles && chartStyles.colorTheme ? chartStyles.colorTheme : { name: DEFAULT_PALETTE };

  const tilingAlgorithm =
    treemapOptions && treemapOptions.tilingAlgorithm && !isEmpty(treemapOptions.tilingAlgorithm)
      ? treemapOptions.tilingAlgorithm[0]
      : 'squarify';

  const sortSectorsField = treemapOptions.sort_sectors || DefaultSortSectors;
  const showColorscale = legend.showLegend ?? 'show';

  const areParentFieldsInvalid =
    new Set([...parentFields.map((field) => field.name)]).size !== parentFields.length ||
    parentFields.some(
      (field) => isEmpty(queriedVizData[field.name]) || isEqual(childField.name, field.name)
    );

  if (
    isEmpty(queriedVizData[childField.name]) ||
    isEmpty(queriedVizData[valueField.name]) ||
    areParentFieldsInvalid
  )
    return <EmptyPlaceholder icon={icontype} />;

  const [treemapData, mergedLayout] = useMemo(() => {
    let labelsArray: string[] = [];
    let parentsArray: string[] = [];
    let valuesArray: number[] = [];
    let colorsArray: string[] = [];

    if (parentFields.length === 0) {
      labelsArray = [...queriedVizData[childField.name]];
      parentsArray = [...Array(labelsArray.length).fill('')];
      valuesArray = [...queriedVizData[valueField.name]];
      if (colorField.name === MULTI_COLOR_PALETTE) {
        colorsArray = [
          ...Array(queriedVizData[childField.name].length).fill(colorField.childColor),
        ];
      }
    } else {
      let currentLevel = parentFields.length - 1;
      let lastParentField = {};
      parentFields
        .slice(0)
        .reverse()
        .map((field, i) => {
          const uniqueParents = uniq(queriedVizData[field.name]) as string[];
          labelsArray = [...labelsArray, ...uniqueParents];
          if (i === 0) {
            parentsArray = [...Array(uniqueParents.length).fill('')];
            valuesArray = [...Array(uniqueParents.length).fill(0)];
            colorsArray =
              colorField.name === MULTI_COLOR_PALETTE
                ? [
                    ...Array(uniqueParents.length).fill(
                      colorField.parentColors[currentLevel] ?? '#000000'
                    ),
                  ]
                : [];
          } else {
            const currentParentIndices = uniqueParents.map((parent) =>
              queriedVizData[field.name].findIndex((index) => index === parent)
            );
            const lastParents = currentParentIndices.map(
              (index) => queriedVizData[lastParentField.name][index]
            );
            parentsArray = [...parentsArray, ...lastParents];
            valuesArray = [...valuesArray, ...Array(lastParents.length).fill(0)];
            colorsArray =
              colorField.name === MULTI_COLOR_PALETTE
                ? [
                    ...colorsArray,
                    ...Array(lastParents.length).fill(
                      colorField.parentColors[currentLevel] ?? '#000000'
                    ),
                  ]
                : [];
          }
          currentLevel = currentLevel - 1;
          lastParentField = field;
        });

      labelsArray = [...labelsArray, ...queriedVizData[childField.name]];
      valuesArray = [...valuesArray, ...queriedVizData[valueField.name]];
      parentsArray = [...parentsArray, ...queriedVizData[lastParentField.name]];
      colorsArray =
        colorField.name === MULTI_COLOR_PALETTE
          ? [
              ...colorsArray,
              ...Array(queriedVizData[childField.name].length).fill(colorField.childColor),
            ]
          : [];
    }

    if (colorField.name === SINGLE_COLOR_PALETTE) {
      colorsArray = [...Array(valuesArray.length).fill(colorField.childColor)];
    }

    const markerColors =
      colorField.name === MULTI_COLOR_PALETTE
        ? { colors: colorsArray }
        : ![DEFAULT_PALETTE, SINGLE_COLOR_PALETTE].includes(colorField.name)
        ? {
            colorscale: colorField.name,
            colorbar: {
              len: 1,
            },
            showscale: showColorscale === 'show',
          }
        : {};

    const colorway = colorField.name === SINGLE_COLOR_PALETTE ? colorsArray : {};

    const mapLayout = {
      ...layout,
      ...(layoutConfig.layout && layoutConfig.layout),
      title: panelOptions.title || layoutConfig.layout?.title || '',
      treemapcolorway: colorway,
      margin: PLOT_MARGIN,
    };

    const mapData = [
      {
        type: 'treemap',
        labels: labelsArray,
        parents: parentsArray,
        values: valuesArray,
        hoverinfo: tooltipMode === 'hidden' ? 'none' : tooltipText,
        textinfo: 'label+value+percent parent+percent entry',
        tiling: {
          packing: tilingAlgorithm.value,
        },
        marker: markerColors,
        sort: sortSectorsField === DefaultSortSectors,
      },
    ];

    return [mapData, mapLayout];
  }, [
    queriedVizData,
    childField,
    valueField,
    parentFields,
    colorField,
    tilingAlgorithm,
    layoutConfig,
  ]);

  const mergedConfigs = {
    ...config,
    ...(layoutConfig.config && layoutConfig.config),
  };

  return <Plt data={treemapData} layout={mergedLayout} config={mergedConfigs} />;
};
