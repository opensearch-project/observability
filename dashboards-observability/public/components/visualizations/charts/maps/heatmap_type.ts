/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { HeatMap } from './heatmap';
import { getPlotlySharedConfigs, getPlotlyCategory } from '../shared/shared_configs';
import { LensIconChartPie } from '../../assets/chart_pie';
import { VizDataPanel } from '../../../event_analytics/explorer/visualizations/config_panel/config_panes/default_vis_editor';
import { ConfigEditor } from '../../../event_analytics/explorer/visualizations/config_panel/config_panes/json_editor';
import {
  ConfigValueOptions,
  HeatmapColorPalettePicker,
  ConfigChartOptions,
  PanelItem,
  SingleColorPicker,
  ConfigLegend,
} from '../../../event_analytics/explorer/visualizations/config_panel/config_panes/config_controls';
import {
  COLOR_PALETTES,
  HEATMAP_SINGLE_COLOR,
  HEATMAP_PALETTE_COLOR,
} from '../../../../../common/constants/colors';
import { fetchConfigObject } from '../../../../components/event_analytics/utils/utils';

const sharedConfigs = getPlotlySharedConfigs();
const VIS_CATEGORY = getPlotlyCategory();

export const createMapsVisDefinition = () => ({
  name: 'heatmap',
  type: 'heatmap',
  id: 'heatmap',
  label: 'Heatmap',
  fulllabel: 'Hubble',
  icontype: 'heatmap',
  category: VIS_CATEGORY.BASICS,
  selection: {
    dataLoss: 'nothing',
  },
  icon: LensIconChartPie,
  editorconfig: {
    panelTabs: [
      {
        id: 'data-panel',
        name: 'Style',
        mapTo: 'dataConfig',
        editor: VizDataPanel,
        sections: [
          fetchConfigObject('Tooltip', {
            options: [
              { name: 'All', id: 'all' },
              { name: 'Dim 1', id: 'x' },
              { name: 'Dim 2', id: 'y' },
              { name: 'Metrics', id: 'z' },
            ],
            defaultSelections: [{ name: 'All', id: 'all' }],
          }),
          {
            id: 'legend',
            name: 'Legend',
            editor: ConfigLegend,
            mapTo: 'legend',
            schemas: [
              {
                name: 'Show colorscale',
                mapTo: 'showLegend',
                component: null,
                props: {
                  options: [
                    { name: 'Show', id: 'show' },
                    { name: 'Hidden', id: 'hidden' },
                  ],
                  defaultSelections: [{ name: 'Show', id: 'show' }],
                },
              },
            ],
          },
          {
            id: 'chart_styles',
            name: 'Chart styles',
            editor: ConfigChartOptions,
            mapTo: 'chartStyles',
            schemas: [
              {
                name: 'Color mode',
                component: PanelItem,
                mapTo: 'colorMode',
                eleType: 'list',
                isSingleSelection: true,
                options: [
                  { name: 'spectrum', label: 'spectrum', value: 'spectrum' },
                  { name: 'opacity', label: 'opacity', value: 'opacity' },
                ],
                defaultState: [{ name: 'spectrum', label: 'spectrum', value: 'spectrum' }],
                props: {
                  isClearable: false,
                },
              },
              {
                name: 'Scheme',
                component: HeatmapColorPalettePicker,
                mapTo: 'scheme',
                eleType: 'palettePicker',
                options: COLOR_PALETTES.filter((color) => color.type !== 'text'),
                defaultState: HEATMAP_PALETTE_COLOR,
              },
              {
                name: 'Color',
                component: SingleColorPicker,
                mapTo: 'color',
                eleType: 'singleColorPicker',
                defaultState: HEATMAP_SINGLE_COLOR,
              },
            ],
          },
        ],
      },
    ],
  },
  visconfig: {
    layout: {
      ...sharedConfigs.layout,
      plot_bgcolor: 'rgba(0, 0, 0, 0)',
      paper_bgcolor: 'rgba(0, 0, 0, 0)',
      margin: { left: 60 },
    },
    config: {
      ...sharedConfigs.config,
    },
  },
  component: HeatMap,
});
