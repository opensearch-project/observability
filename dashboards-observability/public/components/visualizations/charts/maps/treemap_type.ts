/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { TreeMap } from './treemaps';
import { getPlotlySharedConfigs, getPlotlyCategory } from '../shared/shared_configs';
import { LensIconChartBar } from '../../assets/chart_bar';
import { VizDataPanel } from '../../../event_analytics/explorer/visualizations/config_panel/config_panes/default_vis_editor';
import { ConfigEditor } from '../../../event_analytics/explorer/visualizations/config_panel/config_panes/json_editor';
import {
  ConfigValueOptions,
  ColorPalettePicker,
  ConfigChartOptions,
  ConfigLegend,
} from '../../../event_analytics/explorer/visualizations/config_panel/config_panes/config_controls';
import { DEFAULT_PALETTE, COLOR_PALETTES } from '../../../../../common/constants/colors';
import { ButtonGroupItem } from '../../../../../public/components/event_analytics/explorer/visualizations/config_panel/config_panes/config_controls/config_button_group';
import { DEFAULT_CHART_STYLES } from '../../../../../common/constants/shared';
import { fetchConfigObject } from '../../../../components/event_analytics/utils/utils';

const sharedConfigs = getPlotlySharedConfigs();
const VIS_CATEGORY = getPlotlyCategory();

const { SortSectors } = DEFAULT_CHART_STYLES;

export interface BarTypeParams {}

export const createTreeMapDefinition = (params: BarTypeParams = {}) => ({
  name: 'tree_map',
  type: 'tree_map',
  id: 'tree_map',
  label: 'Tree map',
  fulllabel: 'Tree map',
  selection: {
    dataLoss: 'nothing',
  },
  category: VIS_CATEGORY.BASICS,
  icontype: 'heatmap',
  icon: LensIconChartBar,
  categoryaxis: 'xaxis',
  seriesaxis: 'yaxis',
  orientation: 'v',
  component: TreeMap,
  editorconfig: {
    panelTabs: [
      {
        id: 'data-panel',
        name: 'Style',
        mapTo: 'dataConfig',
        editor: VizDataPanel,
        sections: [
          fetchConfigObject('Tooltip', {
            options: [
              { name: 'All', id: 'all' },
              { name: 'Label', id: 'label' },
              { name: 'Value', id: 'value' },
            ],
            defaultSelections: [{ name: 'All', id: 'all' }],
          }),
          {
            id: 'legend',
            name: 'Legend',
            editor: ConfigLegend,
            mapTo: 'legend',
            schemas: [
              {
                name: 'Show colorscale',
                mapTo: 'showLegend',
                component: null,
                props: {
                  options: [
                    { name: 'Show', id: 'show' },
                    { name: 'Hidden', id: 'hidden' },
                  ],
                  defaultSelections: [{ name: 'Show', id: 'show' }],
                },
              },
            ],
          },
          {
            id: 'treemap_options',
            name: 'Treemap',
            editor: ConfigValueOptions,
            mapTo: 'treemapOptions',
            schemas: [
              {
                name: 'Tiling algorithm',
                isSingleSelection: true,
                component: null,
                mapTo: 'tilingAlgorithm',
                options: [
                  { name: 'Squarify', value: 'squarify' },
                  { name: 'Binary', value: 'binary' },
                  { name: 'Dice', value: 'dice' },
                  { name: 'Slice', value: 'slice' },
                  { name: 'Slice Dice', value: 'slice-dice' },
                  { name: 'Dice Slice', value: 'dice-slice' },
                ],
                defaultState: [{ name: 'Squarify', label: 'Squarify', value: 'squarify' }],
                props: {
                  isClearable: false,
                },
              },
              {
                name: 'Sort Sectors',
                component: ButtonGroupItem,
                mapTo: 'sort_sectors',
                eleType: 'buttons',
                props: {
                  options: [
                    { name: 'Largest to Smallest', id: 'largest_to_smallest' },
                    { name: 'Random', id: 'random' },
                  ],
                  defaultSelections: [{ name: 'Largest to Smallest', id: SortSectors }],
                },
              },
            ],
          },
          {
            id: 'chart_styles',
            name: 'Chart styles',
            editor: ConfigChartOptions,
            mapTo: 'chartStyles',
            schemas: [
              {
                name: 'Color theme',
                isSingleSelection: true,
                component: ColorPalettePicker,
                mapTo: 'colorTheme',
                eleType: 'treemapColorPicker',
                options: COLOR_PALETTES,
                defaultState: { name: DEFAULT_PALETTE },
              },
            ],
          },
        ],
      },
      {
        id: 'style-panel',
        name: 'Layout',
        mapTo: 'layoutConfig',
        editor: ConfigEditor,
        content: [],
      },
    ],
  },
  visconfig: {
    layout: {
      ...sharedConfigs.layout,
    },
    config: {
      ...sharedConfigs.config,
    },
    isUniColor: false,
  },
});
