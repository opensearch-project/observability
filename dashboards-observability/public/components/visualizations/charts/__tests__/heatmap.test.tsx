/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import { waitFor } from '@testing-library/react';
import { HeatMap } from '../maps/heatmap';
import {
  LAYOUT_CONFIG,
  TEST_VISUALIZATIONS_DATA
} from '../../../../../test/event_analytics_constants';

describe('Heatmap component', () => {
  configure({ adapter: new Adapter() });

  it('Renders heatmap component', async () => {
    const wrapper = mount(
      <HeatMap
        visualizations={TEST_VISUALIZATIONS_DATA}
        layout={LAYOUT_CONFIG}
        config={{}}
      />
    );

    wrapper.update();

    await waitFor(() => {
      expect(wrapper).toMatchSnapshot();
    });
  });
});
