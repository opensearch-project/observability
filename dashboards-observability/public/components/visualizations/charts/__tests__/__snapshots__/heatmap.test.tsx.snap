// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Heatmap component Renders heatmap component 1`] = `
<HeatMap
  config={Object {}}
  layout={
    Object {
      "colorway": Array [
        "#8C55A3",
      ],
      "height": 220,
      "margin": Object {
        "b": 15,
        "l": 60,
        "pad": 0,
        "r": 10,
        "t": 30,
      },
      "showlegend": true,
    }
  }
  visualizations={
    Object {
      "data": Object {
        "appData": Object {
          "fromApp": false,
        },
        "defaultAxes": Object {},
        "indexFields": Object {
          "availableFields": Array [
            Object {
              "name": "agent",
              "type": "string",
            },
            Object {
              "name": "bytes",
              "type": "long",
            },
            Object {
              "name": "clientip",
              "type": "ip",
            },
            Object {
              "name": "event",
              "type": "struct",
            },
            Object {
              "name": "extension",
              "type": "string",
            },
            Object {
              "name": "geo",
              "type": "struct",
            },
            Object {
              "name": "host",
              "type": "string",
            },
            Object {
              "name": "index",
              "type": "string",
            },
            Object {
              "name": "ip",
              "type": "ip",
            },
            Object {
              "name": "machine",
              "type": "struct",
            },
            Object {
              "name": "memory",
              "type": "double",
            },
            Object {
              "name": "message",
              "type": "string",
            },
            Object {
              "name": "phpmemory",
              "type": "long",
            },
            Object {
              "name": "referer",
              "type": "string",
            },
            Object {
              "name": "request",
              "type": "string",
            },
            Object {
              "name": "response",
              "type": "string",
            },
            Object {
              "name": "tags",
              "type": "string",
            },
            Object {
              "name": "timestamp",
              "type": "timestamp",
            },
            Object {
              "name": "url",
              "type": "string",
            },
            Object {
              "name": "utc_time",
              "type": "timestamp",
            },
          ],
          "queriedFields": Array [
            Object {
              "name": "double_per_ip_bytes",
              "type": "long",
            },
            Object {
              "name": "host",
              "type": "text",
            },
            Object {
              "name": "ip_count",
              "type": "integer",
            },
            Object {
              "name": "per_ip_bytes",
              "type": "long",
            },
            Object {
              "name": "resp_code",
              "type": "text",
            },
            Object {
              "name": "sum_bytes",
              "type": "long",
            },
          ],
          "selectedFields": Array [],
          "unselectedFields": Array [],
        },
        "query": Object {},
        "rawVizData": Object {
          "data": Object {
            "count()": Array [
              154,
              1753,
              116,
              468,
              1964,
              219,
            ],
            "tags": Array [
              "error",
              "info",
              "login",
              "security",
              "success",
              "warning",
            ],
          },
          "jsonData": Array [
            Object {
              "count()": 154,
              "tags": "error",
            },
            Object {
              "count()": 1753,
              "tags": "info",
            },
            Object {
              "count()": 116,
              "tags": "login",
            },
            Object {
              "count()": 468,
              "tags": "security",
            },
            Object {
              "count()": 1964,
              "tags": "success",
            },
            Object {
              "count()": 219,
              "tags": "warning",
            },
          ],
          "metadata": Object {
            "fields": Array [
              Object {
                "name": "count()",
                "type": "integer",
              },
              Object {
                "name": "tags",
                "type": "text",
              },
            ],
            "size": 6,
            "status": 200,
          },
        },
        "userConfigs": Object {},
      },
      "vis": Object {
        "category": "Visualizations",
        "categoryAxis": "xaxis",
        "component": [Function],
        "editorConfig": Object {
          "panelTabs": Array [
            Object {
              "editor": [Function],
              "id": "data-panel",
              "mapTo": "dataConfig",
              "name": "Data",
              "sections": Array [
                Object {
                  "editor": [Function],
                  "id": "value_options",
                  "mapTo": "valueOptions",
                  "name": "Value options",
                  "schemas": Array [
                    Object {
                      "component": null,
                      "isSingleSelection": false,
                      "mapTo": "xaxis",
                      "name": "X-axis",
                    },
                    Object {
                      "component": null,
                      "isSingleSelection": false,
                      "mapTo": "yaxis",
                      "name": "Y-axis",
                    },
                  ],
                },
                Object {
                  "editor": [Function],
                  "id": "chart_options",
                  "mapTo": "chartOptions",
                  "name": "Chart options",
                  "schemas": Array [
                    Object {
                      "component": null,
                      "isSingleSelection": true,
                      "mapTo": "orientation",
                      "name": "Orientation",
                      "props": Object {
                        "defaultSelections": Array [
                          Object {
                            "name": "Vertical",
                            "orientationId": "v",
                          },
                        ],
                        "dropdownList": Array [
                          Object {
                            "name": "Vertical",
                            "orientationId": "v",
                          },
                          Object {
                            "name": "Horizontal",
                            "orientationId": "h",
                          },
                        ],
                      },
                    },
                    Object {
                      "component": null,
                      "isSingleSelection": true,
                      "mapTo": "mode",
                      "name": "Mode",
                      "props": Object {
                        "defaultSelections": Array [
                          Object {
                            "modeId": "group",
                            "name": "Group",
                          },
                        ],
                        "dropdownList": Array [
                          Object {
                            "modeId": "group",
                            "name": "Group",
                          },
                          Object {
                            "modeId": "stack",
                            "name": "Stack",
                          },
                        ],
                      },
                    },
                  ],
                },
              ],
            },
            Object {
              "content": Array [],
              "editor": [Function],
              "id": "style-panel",
              "mapTo": "layoutConfig",
              "name": "Layout",
            },
            Object {
              "editor": [Function],
              "id": "availability-panel",
              "mapTo": "availabilityConfig",
              "name": "Availability",
            },
          ],
        },
        "fullLabel": "Bar",
        "icon": [Function],
        "iconType": "visBarVerticalStacked",
        "id": "bar",
        "label": "Bar",
        "name": "bar",
        "orientation": "v",
        "selection": Object {
          "dataLoss": "nothing",
        },
        "seriesAxis": "yaxis",
        "type": "bar",
        "visConfig": Object {
          "config": Object {
            "displaylogo": false,
            "responsive": true,
          },
          "isUniColor": false,
          "layout": Object {
            "height": 500,
            "legend": Object {
              "orientation": "v",
              "traceorder": "normal",
            },
            "margin": Object {
              "b": 30,
              "l": 60,
              "pad": 0,
              "r": 30,
              "t": 50,
            },
            "showlegend": true,
          },
        },
      },
    }
  }
>
  <EmptyPlaceholder
    icon="visBarVerticalStacked"
  >
    <EuiText
      className="lnsChart__empty"
      color="subdued"
      data-test-subj="vizWorkspace__noData"
      size="xs"
      textAlign="center"
    >
      <div
        className="euiText euiText--extraSmall lnsChart__empty"
        data-test-subj="vizWorkspace__noData"
      >
        <EuiTextAlign
          textAlign="center"
        >
          <div
            className="euiTextAlign euiTextAlign--center"
          >
            <EuiTextColor
              color="subdued"
              component="div"
            >
              <div
                className="euiTextColor euiTextColor--subdued"
              >
                <EuiIcon
                  color="subdued"
                  size="l"
                  type="visBarVerticalStacked"
                >
                  <EuiIconEmpty
                    aria-hidden={true}
                    className="euiIcon euiIcon--large euiIcon--subdued euiIcon-isLoading"
                    focusable="false"
                    role="img"
                    style={null}
                  >
                    <svg
                      aria-hidden={true}
                      className="euiIcon euiIcon--large euiIcon--subdued euiIcon-isLoading"
                      focusable="false"
                      height={16}
                      role="img"
                      style={null}
                      viewBox="0 0 16 16"
                      width={16}
                      xmlns="http://www.w3.org/2000/svg"
                    />
                  </EuiIconEmpty>
                </EuiIcon>
                <EuiSpacer
                  size="s"
                >
                  <div
                    className="euiSpacer euiSpacer--s"
                  />
                </EuiSpacer>
                <p>
                  <FormattedMessage
                    defaultMessage="No results found"
                    id="visualization_noData"
                    values={Object {}}
                  >
                    <span>
                      No results found
                    </span>
                  </FormattedMessage>
                </p>
              </div>
            </EuiTextColor>
          </div>
        </EuiTextAlign>
      </div>
    </EuiText>
  </EmptyPlaceholder>
</HeatMap>
`;
